### PR TITLE
PETScWrappers:BlockSparseMatrix: constructor from PETSc Mat

### PR DESCRIPTION
**Pull request by luca-heltai**
Construct deal.II petsc matrices from petsc Mat types.

### CODE REVIEW

**Review by stefanozampini — APPROVED** [review]

**Comment by drwells** [comment]
Like the other one - can we get rid of `assign_petsc_matrix()`?

**Comment by luca-heltai** [comment]
@drwells, the idea behind the `assign_petsc_matrix()` is precisely to replace the internal mat object of an existing deal.II wrapper with a new object from PETSc. The constructor one is already there.

**Comment by drwells** [comment]
Sorry I've been slow to reply - we also have a `PETScWrappers::SparseMatrix(Mat)` ctor which increments the reference count. Are you OK with doing that and a swap instead of the member function?

**Comment by stefanozampini** [comment]
> Sorry I've been slow to reply - we also have a `PETScWrappers::SparseMatrix(Mat)` ctor which increments the reference count. Are you OK with doing that and a swap instead of the member function?

https://github.com/dealii/dealii/pull/14601#discussion_r1060140921

**Comment by stefanozampini** [comment]
@drwells https://github.com/dealii/dealii/issues/14656

File: include/deal.II/lac/petsc_block_sparse_matrix.h [hunk 1]
```diff
@@ -289,6 +294,17 @@ namespace PETScWrappers
...
+      /**
+       * This method assigns the PETSc Mat to the instance of the class.
+       *
+       * Note that the matrix is not copied: instead, the instance of this class
+       * is initialized to use the given matrix. This is useful if you want to
+       * interpret a PETSc Mat object as a deal.II BlockMatrix, and you already
+       * have a BlockMatrix object that you want to use for this purpose.
+       */
+      void
+      assign_petsc_matrix(Mat A);
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:306]: Also, can you explain (like for the constructor) how block sizes are inferred?

File: include/deal.II/lac/petsc_matrix_base.h [hunk 2]
```diff
@@ -344,6 +344,13 @@ namespace PETScWrappers
      */
     virtual ~MatrixBase() override;
 
+    /**
+     * This method assignes the PETSc Mat to the instance of the class.
+     *
+     */
+    void
+    assign_petsc_matrix(Mat A);
```
Comment (bangerth) [inline include/deal.II/lac/petsc_matrix_base.h:352]: Also copy the statement about ownership from the corresponding function in the block matrix case above.

File: include/deal.II/lac/petsc_sparse_matrix.h [hunk 3]
```diff
@@ -397,6 +403,12 @@ namespace PETScWrappers
        */
       SparseMatrix();
 
+      /**
+       * Initialize a SparseMatrix from a PETSc Mat object. Note that we do not
+       * copy the matrix.
+       */
+      explicit SparseMatrix(const Mat &);
```
Comment (bangerth) [inline include/deal.II/lac/petsc_sparse_matrix.h:410]: And here

File: source/lac/petsc_matrix_base.cc [hunk 4]
```diff
@@ -89,14 +89,23 @@ namespace PETScWrappers
     AssertThrow(ierr == 0, ExcPETScError(ierr));
   }
 
+  void
+  MatrixBase::assign_petsc_matrix(Mat A)
+  {
+    AssertThrow(last_action == ::dealii::VectorOperation::unknown,
+                ExcMessage("Cannot assign a new Mat"));
```
Comment (bangerth) [inline source/lac/petsc_matrix_base.cc:96]: Out of curiosity, why not?
```suggestion
    AssertThrow(last_action == ::dealii::VectorOperation::unknown,
                ExcMessage("Cannot assign a new Mat."));
```

File: source/lac/petsc_matrix_base.cc [hunk 5]
```diff
@@ -89,14 +89,23 @@ namespace PETScWrappers
...
+    PetscErrorCode ierr =
+      PetscObjectReference(reinterpret_cast<PetscObject>(A));
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
```
Comment (bangerth) [inline source/lac/petsc_matrix_base.cc:99]:
```suggestion
    const PetscErrorCode ierr =
      PetscObjectReference(reinterpret_cast<PetscObject>(A));
      (void)ierr;
    AssertThrow(ierr == 0, ExcPETScError(ierr));
```

File: source/lac/petsc_matrix_base.cc [hunk 6]
```diff
@@ -89,14 +89,23 @@ namespace PETScWrappers
...
+    destroy_matrix(matrix);
```
Comment (bangerth) [inline source/lac/petsc_matrix_base.cc:100]: Does this also work if `matrix==nullptr`?

File: source/lac/petsc_matrix_base.cc [hunk 7]
```diff
@@ -89,14 +89,23 @@ namespace PETScWrappers
...
+    matrix = A;
```
Comment (bangerth) [inline source/lac/petsc_matrix_base.cc:101]: You're not keeping track of ownership. That means that in the destructor of this class, you'll call `destroy_matrix()` on the so-assigned `matrix` (==A) object, but that then basically means that the current class takes over ownership.

Or does `PetscObjectReference` above increment a use counter, and then `destroy_matrix` really only does something when the reference count becomes zero? If the latter, please add a couple of lines of comment to the effect above.

Comment (stefanozampini) [reply source/lac/petsc_matrix_base.cc:101]: Yes, objects in PETSc are basically shared pointers.  `PetscObjectReference` increments reference count, XXXDestroy decrements it and, if refcount goes to zero, also destroys the inner data. There's no concept of unique ownership.

File: source/lac/petsc_matrix_base.cc (hunk 6 above)
Comment (stefanozampini) [reply source/lac/petsc_matrix_base.cc:100]: yes

File: source/lac/petsc_matrix_base.cc (hunk 5 above)
Comment (stefanozampini) [reply source/lac/petsc_matrix_base.cc:99]: Why this change? should the `(void)` casting be applied basically everywhere in the code and not just here?

File: source/lac/petsc_matrix_base.cc (hunk 4 above)
Comment (stefanozampini) [reply source/lac/petsc_matrix_base.cc:96]: Well, it is a safety check that no operations are in-fly for the previously held matrix.

File: include/deal.II/lac/petsc_block_sparse_matrix.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:306]: I will make it a private function while implementing https://github.com/dealii/dealii/issues/14656. Is that ok for you?

File: include/deal.II/lac/petsc_sparse_matrix.h [hunk 8]
```diff
@@ -75,6 +75,12 @@ namespace PETScWrappers
      */
     SparseMatrix();
 
+    /**
+     * Initialize a Matrix from a PETSc Mat object. Note that we do not copy
+     * the matrix.
+     */
+    explicit SparseMatrix(const Mat &);
```
[the remark this answers is not in the document]
Comment (stefanozampini) [reply include/deal.II/lac/petsc_sparse_matrix.h:82]: Do you want me to replicate the same sentence for each of those constructors? Something like:

`The Mat object is referenced by the newly created instance of the class using PetscObjectReference. This is in line with the PETSc approach to object ownership, which mimicks std::shared_pointer`.

File: include/deal.II/lac/petsc_block_sparse_matrix.h [hunk 9]
```diff
@@ -299,6 +315,12 @@ namespace PETScWrappers
 
     // ------------- inline and template functions -----------------
 
+    inline BlockSparseMatrix::BlockSparseMatrix(const Mat &A)
+      : BlockSparseMatrix()
+    {
+      this->assign_petsc_matrix(A);
+    }
```
[the remark this answers is not in the document]
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:322]: Should I merge  https://github.com/dealii/dealii/pull/14672 in?

[the remark this answers is not in the document]
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:322]: > Don't you also have to somehow set the block sizes?

That is done inside the function.

File: source/lac/petsc_matrix_base.cc (hunk 7 above)
Comment (bangerth) [reply source/lac/petsc_matrix_base.cc:101]: I've used PETSc for nearly 20 years now and had never realized this :-)

File: source/lac/petsc_matrix_base.cc (hunk 5 above)
Comment (bangerth) [reply source/lac/petsc_matrix_base.cc:99]: Oh, I misread the code. It uses `AssertThrow` instead of `Assert`.

`Assert` is optimized out in release mode, leaving the variable `ierr` assigned to but never read, and as a consequence compilers will warn about that. Saying `(void)ierr;` tells the compiler "We know, treat the variable as used." But `AssertThrow` stays in place in both debug and release mode, and so the variable is used in both cases.

In other words, forget about the suggestion.

File: include/deal.II/lac/petsc_block_sparse_matrix.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:306]: Yes!

File: include/deal.II/lac/petsc_block_sparse_matrix.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:322]: If merging #14672 into this patch makes your life simpler, feel free to. If you don't anticipate merge conflicts, then I guess there is no need.

**Comment by bangerth** [comment]
@stefanozampini Separately, let me say how much I appreciate you taking a look at these interfaces! I wrote the bulk of these PETSc interfaces in 2004 or so, knowing very little about PETSc, and little has been changed since then. I really enjoy seeing where you are taking all of this with your understanding of how PETSc works (in particular the nested objects), and it's going to make these interfaces so much better in the long run -- thank you!

File: include/deal.II/lac/petsc_block_sparse_matrix.h [hunk 10]
```diff
@@ -102,6 +102,11 @@ namespace PETScWrappers
        */
       BlockSparseMatrix() = default;
 
+      /**
+       * Create a BlockSparseMatrix with a PETSc Mat
+       */
+      explicit BlockSparseMatrix(const Mat &);
```
[the remark this answers is not in the document]
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:108]: this has been merged with your PR.

**Comment by stefanozampini** [comment]
@bangerth IMHO this MR is ready to be merged. I will improve the documentation with all your suggestions in a follow-up MR

File: include/deal.II/lac/petsc_block_sparse_matrix.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanozampini) [reply include/deal.II/lac/petsc_block_sparse_matrix.h:306]: renamed reinit here https://github.com/dealii/dealii/pull/14481/commits/757f08692cd4996f064546764b66c6d11c31ecc0

File: include/deal.II/lac/petsc_sparse_matrix.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefanozampini) [reply include/deal.II/lac/petsc_sparse_matrix.h:82]: here https://github.com/dealii/dealii/pull/14481/commits/d76f474c345dcee0b12eb28f64fe1ac890b9665f

File: include/deal.II/lac/petsc_sparse_matrix.h (hunk 3 above)
Comment (stefanozampini) [reply include/deal.II/lac/petsc_sparse_matrix.h:410]: https://github.com/dealii/dealii/pull/14481/commits/d76f474c345dcee0b12eb28f64fe1ac890b9665f

**Comment by stefanozampini** [comment]
@bangerth Your comments have been addressed with follow-up commits.

**Comment by stefanozampini** [comment]
@bangerth Can you please approve?

**Review by drwells — APPROVED** [review]

**Comment by stefanozampini** [comment]
@bangerth this is also ready to go

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
My apologies for the delay!
